src/ucp/core/ucp_mm.c

```diff
@@ -809,6 +805,14 @@ ucs_status_t ucp_memh_get_slow(ucp_context_h context, void *address,
             ucp_memh_length(memh), ucs_memory_type_names[mem_type], reg_md_map,
             uct_flags);
 
+    ucs_assertv(mem_type == memh->mem_type, "mem_type=%s got memh->mem_type=%s",
+                ucs_memory_type_names[mem_type],
+                ucs_memory_type_names[memh->mem_type]);
+    if (context->rcache != NULL) {
```

Comment:
BTW, why the assertion is not valid when not using rcache?

Reply:
when rcache is disabled it doesn't align regions

Reply:
doesn't seem right
why we must align only with rcache?

Reply:
my understanding was that by turning off rcache, we're trying to avoid any unexpected quirks it might bring along
take for instance when we're dealing with rcache and registering on multiple MDs we need to make sure regions are aligned just right to match what the MDs need
this ensures that later on rcache lookups can fetch the region properly
however this alignment can sometimes lead to new headaches
so I thought it'd be neat if disabling rcache also turns off this alignment feature
that way if we disable rcache and the issue disappears we'll have a quick way to tell if it was caused by rcache or not

Reply:
but according to UCT API, the registration must use the alignment specified in uct_md_attr_v2_t.reg_alignment - regardless of caching on higher level
maybe this API field needs to be defined in a better way?

Reply:
sure, but [if rcache is disabled we align registration to uct_md_attr_v2_t.reg_alignment for specific MDs later on](https://github.com/openucx/ucx/pull/9757/files#diff-9136600ae0324c21c07ccead535b761ee5a5113c2817d02af656b33ce992ba8dR519) so their won't interfere